### ISSUE TITLE
GZIP HTML responses

### ISSUE THREAD

**Issue by ddeboer** [zotonic/zotonic#1237]

**Comment by CyBeRoni** [comment]
You are not asking for compressed content. Add --compressed.

(Note that Zotonic still does not answer with compressed content in that case, but the above incantation of curl will never see a Content-Encoding header because curl by default does not send an Accept-Encoding header.)

**Comment by CyBeRoni** [comment]
Please keep in mind [BREACH](http://breachattack.com) ([wiki](https://en.wikipedia.org/wiki/BREACH_(security_exploit))) btw.

**Comment by mworrell** [comment]
At the moment we only send generated HTML page using Identity and not gzip.

It could be easily changed, but needs tests to see if the CPU usage is not too high for doing this (gzip is quite expensive in CPU terms). It also needs a decision on using either Content-Encoding or Transfer-Encoding.  Transfer-Encoding might be the better solution if the reply is chunked.

**Comment by ddeboer** [comment]
Apparently, CSS and JS assets _are_ already being gzipped (and served with `Content-Encoding: gzip`). Is this done in controller_file? 

1. Can we use the same solution for at least GZIP’ing JSON responses from controller_api? 
2. Is a zipped version of (concatenated) asset files cached in the filesystem? This will of course be harder for dynamic API responses.

**Comment by mworrell** [comment]
The css/js files are compressed in memory and then cached (in memory).
This is done by the file-entry process.

Dynamic responses can be compressed using the content-encoding options.
Webmachine should perform the compression.

It is good to check if the compression doesn't take more time than sending the longer response would take.

**Comment by ddeboer** [comment]
Right, so the next step would be to add `{"gzip", fun(X) -> zlib:gzip(X) end}` to `controller_api:content_encodings_provided/2` and measure the results.

**Comment by mworrell** [comment]
Actually in this case it would be the `transfer_encodings_provided`.
As it is all about the transfer and not the content itself.
What `TE` header is sent by your client?

**Comment by mworrell** [comment]
For the `controller_api:content_encodings_provided/2` you could return `["identity", "gzip"]`
The Webmachine routines perform the `gzip` call on the content.

**Comment by ddeboer** [comment]
> Actually in this case it would be the transfer_encodings_provided.

Controller_file uses `content_encodings_provided/2`.

**Comment by mworrell** [comment]
Yes, it has to use `Content-Encoding` as the encoding is done on the content, before the transfer.
So for your controller_api you can experiment with both.
